src/nvim/plines.c

```diff
@@ -991,25 +994,43 @@ int64_t win_text_height(win_T *const wp, const linenr_T start_lnum, const int64_
     lnum = lnum_next + 1;
   }
 
-  while (lnum <= end_lnum) {
+  while (lnum <= *end_lnum && height_sum_nofill + height_sum_fill < max) {
     linenr_T lnum_next = lnum;
-    hasFolding(wp, lnum, &lnum, &lnum_next);
-    height_sum_fill += win_get_fill(wp, lnum);
+    folded_end = hasFolding(wp, lnum, &lnum, &lnum_next);
+    fill_end = win_get_fill(wp, lnum);
+    height_sum_fill += fill_end;
     height_cur_nofill = plines_win_nofill(wp, lnum, false);
```

Comment:
There seems to be a naming inconsistency. `height_cur_nofill`, `folded_end` and `fill_end` refer to properties of the name line, but named in different formats. I'd suggest renaming `folded_end` and `fill_end` to `cur_folded` and `cur_fill`, as `end` may be a bit ambiguous.